fuel-asm/src/lib.rs

```diff
@@ -1,26 +1,627 @@
+            | XOR | RET | ALOC | MCL | MCP | MEQ | ECR | K256 | S256 | NOOP | FLAG | ADDI | ANDI | DIVI | EXPI
+            | MODI | MULI | ORI | SLLI | SRLI | SUBI | XORI | JNEI | LB | LW | SB | SW | MCPI | MCLI | GM | MOVI
+            | JNZI | JI | JMP | JNE | CFEI | CFSI | GTF => true,
+            _ => false,
+        }
+    }
+}
+
+// Direct conversions
+
+impl From<u8> for RegId {
+    fn from(u: u8) -> Self {
+        RegId::new(u)
+    }
+}
+
+impl From<u16> for Imm12 {
+    fn from(u: u16) -> Self {
+        Imm12::new(u)
+    }
+}
+
+impl From<u32> for Imm18 {
+    fn from(u: u32) -> Self {
+        Imm18::new(u)
+    }
+}
+
+impl From<u32> for Imm24 {
+    fn from(u: u32) -> Self {
+        Imm24::new(u)
+    }
+}
+
+impl From<RegId> for u8 {
+    fn from(RegId(u): RegId) -> Self {
+        u
+    }
+}
+
+impl From<Imm12> for u16 {
+    fn from(Imm12(u): Imm12) -> Self {
+        u
+    }
+}
+
+impl From<Imm18> for u32 {
+    fn from(Imm18(u): Imm18) -> Self {
+        u
+    }
+}
+
+impl From<Imm24> for u32 {
+    fn from(Imm24(u): Imm24) -> Self {
+        u
+    }
+}
+
+// Lossless, convenience conversions
+
+impl From<RegId> for usize {
+    fn from(r: RegId) -> usize {
+        u8::from(r).into()
+    }
+}
+
+impl From<Imm12> for u32 {
+    fn from(imm: Imm12) -> Self {
+        u16::from(imm).into()
+    }
+}
+
+impl From<Imm12> for u64 {
+    fn from(imm: Imm12) -> Self {
+        u16::from(imm).into()
+    }
+}
+
+impl From<Imm12> for u128 {
+    fn from(imm: Imm12) -> Self {
+        u16::from(imm).into()
+    }
+}
+
+impl From<Imm18> for u64 {
+    fn from(imm: Imm18) -> Self {
+        u32::from(imm).into()
+    }
+}
+
+impl From<Imm18> for u128 {
+    fn from(imm: Imm18) -> Self {
+        u32::from(imm).into()
+    }
+}
+
+impl From<Imm24> for u64 {
+    fn from(imm: Imm24) -> Self {
+        u32::from(imm).into()
+    }
+}
+
+impl From<Imm24> for u128 {
+    fn from(imm: Imm24) -> Self {
+        u32::from(imm).into()
+    }
+}
+
+impl From<Opcode> for u8 {
+    fn from(op: Opcode) -> Self {
+        op as u8
+    }
+}
+
+impl From<Instruction> for RawInstruction {
+    fn from(inst: Instruction) -> Self {
+        RawInstruction::from_be_bytes(inst.into())
+    }
+}
+
+impl core::convert::TryFrom<RawInstruction> for Instruction {
+    type Error = InvalidOpcode;
+    fn try_from(u: RawInstruction) -> Result<Self, Self::Error> {
+        Self::try_from(u.to_be_bytes())
+    }
+}
+
+// Index slices with `RegId`
+
+impl<T> core::ops::Index<RegId> for [T]
+where
+    [T]: core::ops::Index<usize, Output = T>,
+{
+    type Output = T;
+    fn index(&self, ix: RegId) -> &Self::Output {
+        &self[usize::from(ix)]
+    }
+}
+
+impl<T> core::ops::IndexMut<RegId> for [T]
+where
+    [T]: core::ops::IndexMut<usize, Output = T>,
+{
+    fn index_mut(&mut self, ix: RegId) -> &mut Self::Output {
+        &mut self[usize::from(ix)]
+    }
+}
+
+// Collect instructions into bytes or halfwords
+
+#[cfg(feature = "std")]
```

Comment:
since we're using `core::iter` does this need to be `std` only?

Reply:
We don't currently provide `alloc` (alloc-but-not-std) feature. Should we?

Reply:
Might be useful in certain contexts, e.g. wasm integrations with the VM.

Reply:
WASM supports std, though only a subset of it (I believe `Vec` is included in that).

I'm skeptical any lower-level usage would want to use `alloc`, e.g. In my experience if you can't use `std`, generally you're in such a constrained environment that `alloc`'s `Vec` doesn't make sense either, and you're probably better off with `arrayvec` or something along those lines.

Maybe we're better off waiting until we get requests for features for external collections implementations (like `alloc`) and add them as needed?